File: shared/tweet.js
import React, { useState } from 'react';
import {
  Text,
  View,
  Image,
} from 'react-native';
import { Avatar, Chip } from "@rneui/themed";
import { authedFetch } from './api';
import styles from '../shared/styles';

const timestampDisplayOptions = {
  month: 'short',
  day: 'numeric',
  year: 'numeric',
  hour: '2-digit',
  minute: '2-digit'
}

const TweetBadges = ({ like_count, quote_count, retweet_count, reply_count }) => {
  return (
    <View style={{ paddingLeft: 60, flexDirection: 'row', justifyContent: 'flex-start' }}>
      {like_count > 0    && <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#e91e63", marginRight: 15 }} title='Liked' />}
      {/* {quote_count > 0   && <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#009688", marginRight: 15 }} title='Quoted' />} */}
      {retweet_count > 0 && <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#4caf50", marginRight: 15 }} title='Retweeted' />}
      {/* {reply_count > 0   && <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#ff6f00", marginRight: 15 }} title='Replied' />} */}
    </View>
  );
};

const TweetActions = ({ id, authToken, liked: likedProp = false, retweeted: retweetedProp = false }) => {
  const [liked, setLiked]         = useState(likedProp);
  const [retweeted, setRetweeted] = useState(retweetedProp);

  const onLike = async () => {
    const verb = liked ? 'unlike' : 'like';
    const response = await authedFetch(`tweets/${id}/${verb}`, { token: authToken });
    response.success && setLiked(!liked);
  };

  const onRetweet = async () => {
    const verb = retweeted ? 'unretweet' : 'retweet';
    const response = await authedFetch(`tweets/${id}/${verb}`, { token: authToken });
    response.success && setRetweeted(!retweeted);
  };

  return (
    <View style={{ paddingLeft: 60, flexDirection: 'row', justifyContent: 'flex-start' }}>
      <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#e91e63", marginRight: 15 }} title={liked ? 'Unlike' : 'Like'} onPress={onLike} />
      {/* <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#009688", marginRight: 15 }} title='Quote' /> */}
      <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#4caf50", marginRight: 15 }} title={retweeted ? 'Unretweet' : 'Retweet'} onPress={onRetweet} />
      {/* <Chip size='sm' titleStyle={{ fontSize: 10 }} buttonStyle={{ backgroundColor: "#ff6f00", marginRight: 15 }} title='Reply' /> */}
    </View>
  );
};

const Attachment = ({ type, ...data}) => {
  switch (type) {
    case 'photo':
      return <Image source={{ uri: data.url }} style={{ width: 100, height: 100, borderRadius: 5, margin: 5 }} />;
    case 'video':
      return <Image source={{ uri: data.preview_image_url }} style={{ width: 100, height: 100, borderRadius: 5, margin: 5 }} />;
    default:
      return null;
  }
};

const ReTweet = ({
  id,
  liked,
  retweeted,
  currentUser: { id: currentUserId, token: authToken },
  author: { profile_image_url, name, username, id: authorId },
  referenced_tweet: { created_at, text, author: { username: originalAuthor }, attachments = [] },
  metrics: { like_count, reply_count, retweet_count, quote_count }
}) => {
  const createdAt = (new Date(created_at)).toLocaleDateString(undefined, timestampDisplayOptions);

  return (
    <View style={styles.tweet}>
      <View style={styles.tweetAuthor}>
        <Avatar size={50} rounded source={{ uri: profile_image_url }} containerStyle={styles.tweetAuthorImage} />
        <View>
          <View style={{ ...styles.tweetAuthor, marginBottom: 0 }}>
            <Text style={styles.tweetAuthorName}>{name}</Text>
            <Text style={styles.tweetAuthorUsername}>@{username}</Text>
          </View>
          <Text style={styles.timestamp}>
            {createdAt}
          </Text>
        </View>
      </View>
      <View style={{ paddingLeft: 60, marginBottom: 10 }}>
        <Text>RT @{originalAuthor}:</Text>
        <Text></Text>
        <Text>{text}</Text>
        <View style={{ flexDirection: 'row', flexWrap: 'wrap', justifyContent: 'flex-start', width: 250 }}>
          { attachments.map((data) => <Attachment id={data.media_key} {...data} />) }
        </View>
      </View>
      {currentUserId === authorId ?
        <TweetBadges {...{ like_count, quote_count, retweet_count, reply_count }} /> :
        <TweetActions {...{ id, liked, retweeted, authToken }} />}
    </View>
  );
};

const QuotedTweet = ({
  id,
  liked,
  retweeted,
  currentUser: { id: currentUserId, token: authToken },
  author: { profile_image_url, name, username, id: authorId },
  created_at,
  text,
  referenced_tweet,
  metrics: { like_count, reply_count, retweet_count, quote_count }
}) => {
  const createdAt = (new Date(created_at)).toLocaleDateString(undefined, timestampDisplayOptions);

  return (
    <View style={styles.tweet}>
      <View style={styles.tweetAuthor}>
        <Avatar size={50} rounded source={{ uri: profile_image_url }} containerStyle={styles.tweetAuthorImage} />
        <View>
          <View style={{ ...styles.tweetAuthor, marginBottom: 0 }}>
            <Text style={styles.tweetAuthorName}>{name}</Text>
            <Text style={styles.tweetAuthorUsername}>@{username}</Text>
          </View>
          <Text style={styles.timestamp}>
            {createdAt}
          </Text>
        </View>
      </View>
      <View style={{ paddingLeft: 60 }}>
        <Text>{text}</Text>
      </View>
      <View style={{ paddingLeft: 60, marginBottom: 10, transform: [{ scale: 0.9 }], maxWidth: '100%' }}>
        <StatusTweet {...referenced_tweet} includeBadges={false} maxNameLength={25} />
      </View>
      {currentUserId === authorId ?
        <TweetBadges {...{ like_count, quote_count, retweet_count, reply_count }} /> :
        <TweetActions {...{ id, liked, retweeted, authToken }} />}
    </View>
  );
};

const ReplyTweet = () => {};

const StatusTweet = ({
  id,
  liked,
  retweeted,
  currentUser: { id: currentUserId, token: authToken  } = {},
  author: { profile_image_url, name, username, id: authorId },
  created_at,
  text,
  attachments = [],
  metrics: { like_count, reply_count, retweet_count, quote_count },
  includeBadges = true,
  maxNameLength = 35,
}) => {
  const createdAt = (new Date(created_at)).toLocaleDateString(undefined, timestampDisplayOptions);

  const fullNameLength = name.length + username.length + 1;
  const truncatedName = fullNameLength > maxNameLength ? `${name.slice(0, maxNameLength - username.length - 2)}...` : name;

  return (
    <View style={[styles.tweet, includeBadges ? {} : { borderColor: '#999', borderWidth: 1, borderRadius: 10 }]}>
      <View style={styles.tweetAuthor}>
        <Avatar size={50} rounded source={{ uri: profile_image_url }} containerStyle={styles.tweetAuthorImage} />
        <View>
          <View style={{ ...styles.tweetAuthor, marginBottom: 0 }}>
            <Text style={styles.tweetAuthorName}>{truncatedName}</Text>
            <Text style={styles.tweetAuthorUsername}>@{username}</Text>
          </View>
          <Text style={styles.timestamp}>
            {createdAt}
          </Text>
        </View>
      </View>
      <View style={{ paddingLeft: 60, marginBottom: 10 }}>
        <Text>{text}</Text>
        <View style={{ flexDirection: 'row', flexWrap: 'wrap', justifyContent: 'flex-start', width: 250 }}>
          { attachments.map((data) => <Attachment id={data.media_key} {...data} />) }
        </View>
      </View>
      {includeBadges && currentUserId === authorId && <TweetBadges {...{ like_count, quote_count, retweet_count, reply_count }} />}
      {includeBadges && currentUserId !== authorId && <TweetActions {...{ id, liked, retweeted, authToken }} />}
    </View>
  );
};

export default function Tweet({ type, ...props}) {
  switch(type) {
    case 'quoted':
      return <QuotedTweet {...props} />;
    case 'replied_to':
      return <StatusTweet {...props} />;
    case 'retweeted':
      return <ReTweet {...props} />;
    default:
      return <StatusTweet {...props} />;
  }
}

File: server/app/models/tweet.rb
class Tweet
  class Metrics
    class << self
      def from_twitter(data)
        new(**data)
      end
    end

    attr_reader :like_count, :retweet_count, :quote_count, :reply_count

    def initialize(like_count: 0, retweet_count: 0, quote_count: 0, reply_count: 0, **_others)
      @like_count    = like_count
      @retweet_count = retweet_count
      @quote_count   = quote_count
      @reply_count   = reply_count
    end
  end

  class Attachment
    class << self
      def from_twitter(data)
        new(**data)
      end
    end

    attr_reader :media_key, :type, :url, :preview_image_url

    def initialize(media_key:, type:, url: nil, preview_image_url: nil, **_others)
      @media_key         = media_key
      @type              = type
      @url               = url
      @preview_image_url = preview_image_url
    end

    def to_partial_path
      "tweets/attachment"
    end
  end

  class Author
    class << self
      def from_twitter(data)
        new(**data)
      end
    end

    attr_reader :id, :username, :name, :profile_image_url, :description, :pinned_tweet, :following, :in_feed, :muted, :blocked

    def initialize(id:, username:, name:, profile_image_url:, description: nil, pinned_tweet: nil, following: false, in_feed: false, muted: false, blocked: false, **_others)
      @id                = id
      @username          = username
      @name              = name
      @profile_image_url = profile_image_url
      @description       = description&.gsub("\u2028", "\n")
      @pinned_tweet      = pinned_tweet
      @following         = following
      @in_feed           = in_feed
      @muted             = muted
      @blocked           = blocked
    end

    def to_partial_path
      "authors/author"
    end
  end

  class << self
    def from_response(response, likes: [], retweets: [])
      Array.wrap(response[:data]).map do |tweet|
        if (referenced_tweet = Array.wrap(response.dig(:includes, :tweets)).find { |rt| tweet.dig(:referenced_tweets, 0, :id) == rt[:id] })
          referenced_tweet_attachments = referenced_tweet.dig(:attachments, :media_keys) || []
          if referenced_tweet_attachments.any?
            referenced_tweet_attachments = referenced_tweet_attachments.map do |media_key|
              (response.dig(:includes, :media) || []).find { |media| media[:media_key] == media_key }
            end.compact
          end

          referenced_tweet = referenced_tweet.merge(
            type: referenced_tweet[:referenced_tweets]&.any? ? tweet.dig(:referenced_tweets, 0, :type) : 'tweet',
            author: response.dig(:includes, :users).find { |user| user[:id] == referenced_tweet[:author_id] },
            attachments: referenced_tweet_attachments,
            liked: likes.include?(referenced_tweet[:id]),
            retweeted: retweets.include?(referenced_tweet[:id])
          )
        end

        attachments = tweet.dig(:attachments, :media_keys) || []
        if attachments.any?
          attachments = attachments.map do |media_key|
            response.dig(:includes, :media).find { |media| media[:media_key] == media_key }
          end.compact
        end

        from_twitter(
          **tweet.slice(:id, :conversation_id, :text, :created_at, :public_metrics),
          type: tweet[:referenced_tweets]&.any? ? tweet.dig(:referenced_tweets, 0, :type) : 'tweet',
          author: response.dig(:includes, :users).find { |user| user[:id] == tweet[:author_id] },
          referenced_tweet: referenced_tweet,
          attachments: attachments,
          liked: likes.include?(tweet[:id]),
          retweeted: retweets.include?(tweet[:id])
        )
      end
    end

    def from_twitter(data)
      new(
        **data.merge(
          author: Author.from_twitter(data[:author]),
          metrics: Metrics.from_twitter(data.dig(:public_metrics)),
          referenced_tweet: data.dig(:referenced_tweet) ? from_twitter(data.dig(:referenced_tweet)) : nil,
          attachments: (data.dig(:attachments) || []).map { |attachment| Attachment.from_twitter(attachment) }
        )
      )
    end
  end

  attr_reader :id, :conversation_id, :text, :author, :created_at, :referenced_tweet, :type, :metrics, :attachments, :liked, :retweeted

  def initialize(
    id:,
    text:,
    author:,
    created_at:,
    type:,
    metrics:,
    conversation_id: nil,
    referenced_tweet: nil,
    attachments: [],
    liked: false,
    retweeted: false,
    **_others
  )
    @id               = id
    @conversation_id  = conversation_id
    @text             = text.gsub("\u2028", "\n")
    @author           = author
    @created_at       = created_at
    @type             = type
    @metrics          = metrics
    @referenced_tweet = referenced_tweet
    @attachments      = attachments
    @liked            = liked
    @retweeted        = retweeted
  end

  def quote?
    type == "quoted"
  end

  def reply?
    type == "replied"
  end

  def retweet?
    referenced_tweet.present? && !reply? && !quote?
  end

  def to_partial_path
    "tweets/tweet"
  end
end
